### PR TITLE
Fixes autocomplete when using sudo

### PR DESCRIPTION
**Pull request by alesito85**
# Description

This PR addresses issue #2047 in order to enable autocomplete functionality when using sudo for executing commands. I'e done a couple of auxiliary checks such as ignoring whitespace and the last pipe in order to determine the last command.

# User-Facing Changes

The only user facing change should be the autocomplete working.

# Tests + Formatting

All tests and formatting pass.

# Screenshots

image

# Suggestions welcome

I still don't know the in's and out's if nushell very well, any suggestions for improvements are welcome.

### CODE REVIEW

**Comment by alesito85** [comment]
@rgwood I'm not really sure. what to look for in the failed CI since everything passes locally. Could this be a temporary issue or something that actually needs a fix?

**Comment by sholderbach** [comment]
I think you came across the flakiness of the plugin tests on the Mac CI runners. Restarted them, should typically resolve.

File: crates/nu-protocol/src/engine/engine_state.rs
Comment (sholderbach) [inline crates/nu-protocol/src/engine/engine_state.rs:2475]: Maybe I am wrong but this feels like two independent tests in one. 
The test name says it tests your string trimming helper but it also contains code specific to the `sudo` goal. For the former it feels like overtesting but that might not be the case for the latter.

File: crates/nu-protocol/src/engine/engine_state.rs
Comment (sholderbach) [inline crates/nu-protocol/src/engine/engine_state.rs:1569]: No magic numbers please.
```suggestion
            let last_pipe_pos = contents.iter().rev().position(|x| x == b'|');
```

File: crates/nu-protocol/src/engine/engine_state.rs
Comment (sholderbach) [inline crates/nu-protocol/src/engine/engine_state.rs:1573]: In my mappy-foldy-iterator view of the Rust world that reads like `.map(...).unwrap_or_default()`, but that is a style question how comfortable you are with reading either style.

File: crates/nu-cli/src/completions/completer.rs
Comment (sholderbach) [inline crates/nu-cli/src/completions/completer.rs:147]: This `.expect()` feels like danger zone just browsing through the patch. We have still a bunch of panics to track down with completions so I don't feel like that is the right way to go about this line.

Something along the lines of `.filter()`/`.map()`/`Option::contains()` would get you there without trying to check and promise you can unwrap it.
(Probably needs some `Deref` or else to get the types right)

```
                            let is_sudo_command =spans.first().filter(|content| content == "sudo").is_some();
```

**Review by sholderbach — COMMENTED** [review]
Thank you very much for tackling this and figuring out some way to get you there.

I am not that up to speed on how the completion parsing and generation works at the moment, so I can not weigh in too much on that. But I have some concerns about the implementation that would make the code harder to maintain in the future.

To convince us to take this change it would probably also be beneficial to have a full test of a completion with `sudo` prefixed that includes some potential edge cases to make sure we don't break unexpectedly.

**Comment by alesito85** [comment]
Thank you for taking your time. I've applied fixes according to your comments, they make sense to me and I see I'll have to check more of what Rust has already implemented to not do double work where applicable.

I still have to add the test, which I expect to do shortly.

**Comment by alesito85** [comment]
@sholderbach I believe I've applied all the changes, including the test. Hopefully it captures what you meant.

**Comment by sholderbach** [comment]
Thanks for updating the code and adding the additional independent tests!
I think this addresses what I asked for.

I would hold off on merging for a moment as @kubouch is currently fighting with the completion logic a bit to track down a nasty panic or logic bug (#7779 #8102). We certainly need to reign in a bunch of kludgy special cases that already exist and make everything a bit brittle. If we get this cleared it is probably easier to say how your fix fits in.
So we probably have to wait until after the imminent release.

This maybe also gives us some time to clarify how sudo should behave in regards to internal and external commands (referring to your #8099)

Thanks for being responsive and eager to help out with nushell!

**Comment by kubouch** [comment]
I'm OK with it, I was tracking mostly bugs related to aliases, this seems unrelated.

I would move `find_non_whitespace_index()` and `is_sudo()` away from the EngineState because the first does not reference EngineState at all and I'm not sure we want to hard-code sudo detection into the EngineState.

Another change I would do is to rename `is_sudo()` to something more generic, like `is_passthrough_command()` because tools like [`bear`](https://github.com/rizsotto/Bear) or `gdb` would benefit from the same autocomplete support as `sudo`. We can keep "sudo" hardcoded there for now and consider adding support for more commands later. When adding support for more command, we don't want to hard-code every single command into Nushell, so one way to do it would be to have an environment variable with a list of commands (populated by `sudo` by default) that the completer would look up.

**Comment by alesito85** [comment]
Excellent suggestions, thank you. I'll take care of them tomorrow I think.

File: crates/nu-protocol/src/engine/engine_state.rs
Comment (sholderbach) [inline crates/nu-protocol/src/engine/engine_state.rs:1012]: Can we get away here without cloning?

Comment (kubouch) [reply crates/nu-protocol/src/engine/engine_state.rs:1012]: Agreed, you'll be copying the entire source ever read by Nushell just to find sudo, that seems like an overkill. Most likely you can return `&[u8]`.

**Comment by kubouch** [comment]
Instead of utils, I would put the two functions into the command_completer.rs because they're used only for the completions.

The logic for detecting the sudo is also a bit gnarly. The completer has access to the parsed result so there is no need to iterate bytes, you should be able to check for Expr::ExternalCall and see if it matches sudo and go from there. Or something like that. But I think as long as it works and does not break anything, it should be OK to merge as well.

**Comment by alesito85** [comment]
This will benefit from some squashing before merging. Let me know if it's alright as it is.

**Comment by kubouch** [comment]
We won't merge it at least for the next 1-2 days because we're doing the release, so if you want to keep hacking on it, go ahead. In general, looking up the bytes manually seems a bit unsound to me, I'd rather use the parsed output.

One more thing, though. You're using direct array access with `[]` which can panic. In Nushell, if something can panic, it _will_ panic at some point. It would be better to use `.get()` or something similar that doesn't panic.

**Comment by alesito85** [comment]
No worries, I'm not in rush for having this merged, just saying when you feel it's ready, I'll also also squash the changes.

I've replaced [..] with .get, thanks for pointing it out to me.

I think I'm only using bytes where I don't have the full string available so that can probably stay like that? I'm perfectly fine with adapting the code if you point it out where you think I should do it differently.

**Review by sholderbach — APPROVED** [review]
OK seems good now!

**Comment by myrkvi** [comment]
This doesn't seem to allow completions to complete local paths (i.e. programs that are not in PATH).

#### Example:
You have a script in a folder that requires root privileges to run. It is not in the PATH environment variable.
If you are currently in your home directory, and the script is in `~/scripts/do_admin_stuff.nu`, you could invoke it as `sudo scripts/do_admin_stuff.nu` or `sudo ./scripts/do_admin_stuff.nu`. Trying to complete either of these paths results in the message "NO RECORDS FOUND". Trying to define an extern to override this functionality also does not work.

**Pull request merged**
